File: go/arrays-and-hashing/two_sum_test.go
package arrays_and_hashing

import (
	"testing"
	"time"
)

func TestTwoSum(t *testing.T) {
	tests := []struct {
		name   string
		nums   []int
		target int
		want   []int
	}{
		{
			name:   "basic case",
			nums:   []int{2, 7, 11, 15},
			target: 9,
			want:   []int{0, 1},
		},
		{
			name:   "different positions",
			nums:   []int{3, 2, 4},
			target: 6,
			want:   []int{1, 2},
		},
		{
			name:   "same value different indices",
			nums:   []int{3, 3},
			target: 6,
			want:   []int{0, 1},
		},
		{
			name:   "negative numbers",
			nums:   []int{-1, -2, -3, -4, -5},
			target: -8,
			want:   []int{2, 4},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := TwoSum(tt.nums, tt.target)
			if len(got) != len(tt.want) {
				t.Errorf("TwoSum() = %v, want %v", got, tt.want)
				return
			}
			for i := range got {
				if got[i] != tt.want[i] {
					t.Errorf("TwoSum() = %v, want %v", got, tt.want)
					return
				}
			}
		})
	}
}

func BenchmarkTwoSum1K(b *testing.B) {
	nums := make([]int, 1000)
	for i := range nums {
		nums[i] = i
	}
	target := 1997

	start := time.Now()
	result := TwoSum(nums, target)
	elapsed := time.Since(start)

	b.Logf("[1K elements] Execution time: %.3fms", float64(elapsed.Microseconds())/1000)

	if result[0] != 998 || result[1] != 999 {
		b.Errorf("Expected [998, 999], got %v", result)
	}
}

func BenchmarkTwoSum10K(b *testing.B) {
	nums := make([]int, 10000)
	for i := range nums {
		nums[i] = i
	}
	target := 19997

	start := time.Now()
	result := TwoSum(nums, target)
	elapsed := time.Since(start)

	b.Logf("[10K elements] Execution time: %.3fms", float64(elapsed.Microseconds())/1000)

	if result[0] != 9998 || result[1] != 9999 {
		b.Errorf("Expected [9998, 9999], got %v", result)
	}
}

func BenchmarkTwoSum100K(b *testing.B) {
	nums := make([]int, 100000)
	for i := range nums {
		nums[i] = i
	}
	target := 199997

	start := time.Now()
	result := TwoSum(nums, target)
	elapsed := time.Since(start)

	b.Logf("[100K elements] Execution time: %.3fms", float64(elapsed.Microseconds())/1000)

	if result[0] != 99998 || result[1] != 99999 {
		b.Errorf("Expected [99998, 99999], got %v", result)
	}
}

File: typescript/arrays-and-hashing/two-sum.test.ts
import { twoSum } from './two-sum';

describe('twoSum', () => {
    test('should return indices [0, 1] for nums=[2,7,11,15] and target=9', () => {
        const nums = [2, 7, 11, 15];
        const target = 9;
        const result = twoSum(nums, target);
        expect(result).toEqual([0, 1]);
    });

    test('should return indices [1, 2] for nums=[3,2,4] and target=6', () => {
        const nums = [3, 2, 4];
        const target = 6;
        const result = twoSum(nums, target);
        expect(result).toEqual([1, 2]);
    });

    test('should return indices [0, 1] for nums=[3,3] and target=6', () => {
        const nums = [3, 3];
        const target = 6;
        const result = twoSum(nums, target);
        expect(result).toEqual([0, 1]);
    });

    test('should handle negative numbers', () => {
        const nums = [-1, -2, -3, -4, -5];
        const target = -8;
        const result = twoSum(nums, target);
        expect(result).toEqual([2, 4]);
    });
});

describe('twoSum - Performance Benchmarks', () => {
    test('benchmark with 1,000 elements', () => {
        const nums = Array.from({ length: 1000 }, (_, i) => i);
        const target = 1997;
        
        const start = performance.now();
        const result = twoSum(nums, target);
        const end = performance.now();
        
        console.log(`[1K elements] Execution time: ${(end - start).toFixed(3)}ms`);
        expect(result).toEqual([998, 999]);
    });

    test('benchmark with 10,000 elements', () => {
        const nums = Array.from({ length: 10000 }, (_, i) => i);
        const target = 19997;
        
        const start = performance.now();
        const result = twoSum(nums, target);
        const end = performance.now();
        
        console.log(`[10K elements] Execution time: ${(end - start).toFixed(3)}ms`);
        expect(result).toEqual([9998, 9999]);
    });

    test('benchmark with 100,000 elements', () => {
        const nums = Array.from({ length: 100000 }, (_, i) => i);
        const target = 199997;
        
        const start = performance.now();
        const result = twoSum(nums, target);
        const end = performance.now();
        
        console.log(`[100K elements] Execution time: ${(end - start).toFixed(3)}ms`);
        expect(result).toEqual([99998, 99999]);
    });

    test('benchmark worst case - target at end', () => {
        const nums = Array.from({ length: 10000 }, (_, i) => i);
        const target = 19997; // Last two elements sum
        
        const start = performance.now();
        const result = twoSum(nums, target);
        const end = performance.now();
        
        console.log(`[Worst case - 10K] Execution time: ${(end - start).toFixed(3)}ms`);
        expect(result).toEqual([9998, 9999]);
    });
});